### DISCUSSION TITLE
Servos and shutters

### DISCUSSION THREAD

**Discussion by jpadie** [arendst/Tasmota#19113] [Q&A]
I am using an ESP32 (13.0.0.) as a servo controller for a cat-feeder.  
I've been able to get this working reliably using pwm commands.  However I think the lack of slow start prejudices the motor as I have has two burn out.  
I'm trying to get the shuttermode to work but whatever the values I use for frequency and pwmrange I cannot reliably control the position (most of the time a position command seems to move the servo to an angle and then move it back to another angle).
I have followed the guides as best I can and used the values provided for an MG90 (which I am using an MG966 but the values for frequency, duty cycle and min/max pulse width are the same) but, as said, no reliable results.  

I may be confusing myself, of course.  e.g. taking the value for the MG90 as a paradigm; the site reports that 

> Servo MG90 (180°): pulse width range 544 - 2400ms, ShutterPwmRange 109, 480

however the linked data sheet says the pulse duration is 20ms and the durations for min and max angles are 1 and 2 ms respectively.  How does this correlate to a PW range of 544-2400ms? 

and if the correct position is known from vanilla PWM commands is there a method of converting these into values that can be used in the shutter configuration?  

thanks in advance.

**Comment by Jason2866** [comment]
Which pwm frequency do you use? Usual servos like the MG90 are driven with 50Hz. 200Hz can work fine, but when problems occur the standard servo frequency is worth a try.
Probably this "pulse width range 544 - 2400ms" is just wrong in the documentation. @stefanbode 
```
Position "0" (1.5 ms pulse) is middle, "90" (~2 ms pulse) is all the way to the right, "-90" (~1 ms pulse) is all the way to the left.
```

Digital servos are best driven with 333Hz.

Servos do not need any sort of slow start. The are made for fixed frequency which is pwm modulated.

**Reply by jpadie** [reply]
thanks for the reply Jason. 
I've tried with frequency set at 50Hz (per the datasheet) and at 200hz per the wiki.  

I think you are right about the wiki text being wrong.  I suspect it should be "544-2400us" not ms.  

> Servos do not need any sort of slow start. The are made for fixed frequency which is pwm modulated.

noted.  then I am at a loss as to why two have burned out with Tasmota but the others in the estate work fine.  Perhaps it is a question of tasmota not releasing the PWM after setting it (pwm<x> 0).  I see there is a compile time definition that sets the  pwm to 0 (SHUTTER_CLEAR_PWM_ONSTOP).  perhaps that will assist.   Perhaps it can be done through berry is there is an event that is fired after the move is complete.

**Reply by stefanbode** [reply]
Hi. Yes is for sure microseconds and not milli. I updated the documentation already. The compiler option is for disable the PWM after movement. Maybe there is to much load on the servo and here it overheats. You can use the first GPIO for the on/off. This should work with a transistor to switch the 5V. You can see this in the picture v522 in the documentation with a relay. 


I made all development with these small servos you use for airplanes and ships in miniature. Servos support acceleration at start and stop with shutter motor delay.

**Reply by stefanbode** [reply]
You can fire a PWM=0 or dimmer=0 after movement to get the same effect. There is a rule trigger shutter#moved that can do it.

**Reply by Jason2866** [reply]
Servos should not have a to hold a mechanical position after movement. The are built to move something in a position (with the Nm the are designed) after that no force should be needed anymore. If this is the case the servo motor is a always powered and will burn out.
So switching off the PWM signal is not needed when the servo is used correctly.
Servo electronic will power the motor only when the targeted position is not anymore the targeted one.
Typical example is for steering a model car. Servo holds the wheels in the wanted direction, against other forces. Servo motor cools down when driving straight forward. No force on wheels, servo motor is not powered.

**Reply by Jason2866** [reply]
So check the current the servo needs goes done to a few mA after reaching the target position.
If not you have to rethink your mechanical design. Every servo will get damaged when motor is needed to be powered to hold the position.

**Reply by jpadie** [reply]
Thanks Jason

Today has killed three of the four servos i have spare in the lab so I have stopped for now.   The cats need feeding so tomorrow I will go to a plain c++ sketch which I know works and come back to a tasmota implementation in September.

**Reply by stefanbode** [reply]
Would be great if you can share the sketch so we an in the meantime fix the code.
